# Projeto SQL


Banco de dados de um dos serviços concorrentes nesse mercado que contém dados sobre livros, editoras, autores, e classificação de clientes e avaliação de livros. Essa informação será usada para gerar uma proposição válida para o novo produto.

# 1. Descrição dos dados:
1. __Arquivo 1: contém informações sobre os livros - *(books — livros):*__

    * *book_id* — identificador do livro
    * *author_id* — identificador do autor
    * *title* — título
    * *num_pages* — número de páginas
    * *publication_date* — data de publicação
    * *publisher_id* — identificador da editora

2. __Arquivo 2: contém informações sobre os autores - *(authors — autores):*__

    * *author_id* — identificador do autor
    * *author* — autor

3. __Arquivo 3: contém informações sobre as editoras - *(publishers — editoras):*__

    * *publisher_id* — identificador da editora
    * *publisher* — editora

4. __Arquivo 4: contém informações sobre as classificação dos usuários - *(ratings — classificações):*__

    * *rating_id* — identificador da classificação
    * *book_id* — identificador do livro
    * *username* — o nome do usuário que avaliou o livro
    * *rating* — classificação
    
5. __Arquivo 5: contém informações sobre as dados sobre revisão dos clientes - *(reviews — avaliação):*__

    * *review_id* — identificador da revisão
    * *book_id* — identificador do livro
    * *username* — o nome do usuário que revisou o livro
    * *text* — o texto da revisão
    
    

# 2. Diagrama dos Dados 

<img src ="moved_Untitled_-_2020-07-02T142019.920_1593688954.png">

# 3. Tarefa

   1. Encontre o número de livros lançados depois de 1 de janeiro de 2000.
   2. Encontre o número de avaliações e a classificação média para cada livro.
   3. Identifique a editora que lançou o maior número de livros com mais de 50 páginas (isso vai ajudar você a excluir brochuras e publicações parecidas da sua análise).
   4. Identifique o autor com a média mais alta classificação de livros: olhe apenas para livros com pelo menos 50 classificações.
   5. Encontre o número médio de avaliações entre usuários que classificaram mais do que 50 livros.


# 4. Instruções para completar a tarefa

   1. Descreva os objetivos do estudo.
   2. Estude as tabelas (imprima as primeiras linhas).
   3. Faça uma consulta SQL para cada uma das tarefas.
   4. Enuncie os resultados de cada consulta no Caderno.

# 5. Descreva suas conclusões para cada uma das tarefas.

___________________

## Início

### Carregando as bibliotecas e acesso ao banco de dados

In [1]:
# importar bibliotecas
import pandas as pd
from sqlalchemy import create_engine

In [2]:
## Acessando o banco de dados
db_config = {'user': 'practicum_student', # nome de usuário
'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # senha
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # porta de conexão
'db': 'data-analyst-final-project-db'} # o nome do banco de dados
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Conjunto de dados books

In [3]:
#Acessando a tabela 'books' 
query = '''
SELECT* 
    FROM books
'''
books = pd.io.sql.read_sql(query, con = engine)
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
#Resumo estatístico do conjunto de dados
books.describe()

book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

In [5]:
#Informações do conjunto de dados
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Conjunto de dados 'authors'

In [6]:
#Acessando a tabela 'authors'
query = '''
SELECT* 
    FROM authors
'''
authors = pd.io.sql.read_sql(query, con = engine)
authors

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [7]:
#Resumo estatístico do conjunto de dados
authors.describe()

author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

In [8]:
#Informações do conjunto de dados
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Pelas verificações acima, percebe-se a consistência dos dados. Não há valores nulos ou valores ausentes.

### Conjunto de dados 'publishers'

In [9]:
#Acessando a tabela 'publishers'
query = '''
SELECT* 
    FROM publishers
'''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [10]:
#Resumo estatístico do conjunto de dados
publishers.describe()

publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

In [11]:
#Informações do conjunto de dados
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Pelas verificações acima, percebe-se a consistência dos dados. Não há valores nulos ou valores ausentes.

### Conjunto de dados 'ratings'

In [12]:
#Acessando a tabela 'ratings'
query = '''
SELECT* 
    FROM ratings
'''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [13]:
#Resumo estatístico do conjunto de dados
ratings.describe()

rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

In [14]:
#Informações do conjunto de dados
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Pelas verificações acima, percebe-se a consistência dos dados. Não há valores nulos ou valores ausentes.

### Conjunto de dados 'reviews'

In [15]:
#Acessando a tabela 'reviews'
query = '''
SELECT* 
    FROM reviews
'''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [16]:
#Resumo estatístico do conjunto de dados
reviews.describe()

review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

In [17]:
#Informações do conjunto de dados
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Pelas verificações acima, percebe-se a consistência dos dados. Não há valores nulos ou valores ausentes.

## Consultas em SQL

### Número de livros lançados após 1 de janeiro de 2000

In [18]:
#Conjunto de dados
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [19]:
#Realizando a consulta
query = '''
SELECT
    SUM(z.livros) AS "qtde_livros"
FROM    (SELECT
        COUNT(t.title) AS "livros"
        FROM  (SELECT
                books.title
                FROM books
                WHERE publication_date > '2000-01-01') AS t
                GROUP BY
                    t.title) AS z
'''
books_public = pd.io.sql.read_sql(query, con = engine)
books_public

qtde_livros
0        819.0

Acima nota-se a tabela gerada com as informações dos livros publicados a partir de 01-01-2000, onde a data de publicação vai de 05-01-2000 até 31-03-2020.

### Número de avaliações e a classificação média para cada livro.

In [20]:
#Conjunto de dados
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [21]:
#Conjunto de dados
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [22]:
#Conjunto de dados
ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [23]:
#Realizando a consulta
query='''
SELECT
    books.book_id,
    books.title,
    (SELECT 
        COUNT(reviews.book_id)
    FROM reviews
    WHERE reviews.book_id = books.book_id) AS qtd_avaliações,
    (SELECT 
        AVG(ratings.rating)
    FROM ratings
    WHERE ratings.book_id = books.book_id) AS class_media
FROM books
ORDER BY
    qtd_avaliações DESC;
'''

books_avg_rating = pd.io.sql.read_sql(query,con= engine)
books_avg_rating

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     qtd_avaliações  class_media  
0                 7     3.662500  
1                 6     3.977273  
2                 6     4.206897  
3                 6     4.414634  
4                 6     4.081081  
..              ...          ...  
995               0     3.666667  
996               0     3.000000  
997               0     5.000000  
998               0     4.000000  
999               0     4.000000  

[1000 rows x 4 columns]

In [24]:
#Descrição do conjunto de dados
books_avg_rating.describe()

book_id  qtd_avaliações  class_media
count  1000.000000     1000.000000  1000.000000
mean    500.500000        2.793000     3.898973
std     288.819436        1.074852     0.562376
min       1.000000        0.000000     1.500000
25%     250.750000        2.000000     3.500000
50%     500.500000        3.000000     4.000000
75%     750.250000        3.000000     4.333333
max    1000.000000        7.000000     5.000000

In [25]:
#Agrupando por 'qtde_avaliações'
books_avg_rating.groupby('qtd_avaliações').count()

book_id  title  class_media
qtd_avaliações                             
0                     6      6            6
1                    37     37           37
2                   444    444          444
3                   277    277          277
4                   160    160          160
5                    60     60           60
6                    15     15           15
7                     1      1            1

Pelas tabelas geradas acima, nota-se a distribuição das quantidades de avaliações e a classificação média, onde é possivel verificar que o range vai de 0 a 7 em quantidade de avaliações , onde cerca de 72% dos livros são avaliados entre 2  e 3 vezes, 6 livros não tiveram avaliação e apenas 1 livros foi avaliado 7 vezes

### Editora que lançou o maior número de livros com mais de 50 páginas.

In [26]:
#Conjunto de dados
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [27]:
#Conjunto de dados
publishers

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [28]:
#Realizando a consulta
query ='''
SELECT
    publisher,
    COUNT(*) AS "count_publishers"
FROM (
SELECT
    p.publisher
FROM
    books b
JOIN
    publishers p USING (publisher_id)
WHERE
    b.num_pages > 50) AS PU
GROUP BY
    publisher
ORDER BY
    count_publishers DESC
LIMIT 1;
'''

books_publ = pd.io.sql.read_sql(query,con= engine)
books_publ.head(10)

publisher  count_publishers
0  Penguin Books                42

In [29]:
books_publ['count_publishers'].sum()

42

Pelas tabelas geradas acima, verifica-se a distribuição dos livros por editora, onde é possivel notar que as TOP10 editoras ( das 340 editoras listadas) respondem por cerca de 22% de toda a publicação do período.

### Autor com a média mais alta de classificação de livros: livros com pelo menos 50 classificações.

In [30]:
#Conjunto de dados
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [31]:
#Conjunto de dados
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [32]:
#Conjunto de dados
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [33]:
#Realizando a consulta
query ='''
SELECT
    b.author_id,
    AVG(t.avg_rating) AS avg_rating
FROM
    books b
INNER JOIN(
SELECT
    r.book_id,
    COUNT(r.book_id) AS "qtde_rating",
    AVG(r.rating) AS "avg_rating"
FROM
    ratings r
GROUP BY
    book_id
HAVING
    COUNT(r.book_id) > 50) t ON t.book_id = b.book_id
GROUP BY
    author_id
ORDER BY
    AVG(t.avg_rating) DESC
'''

books_z = pd.io.sql.read_sql(query,con= engine)
books_z

author_id  avg_rating
0         236    4.283844
1         402    4.264151
2         240    4.258446
3         376    4.192308
4         498    4.080645
5         621    3.901408
6         235    3.825581
7         469    3.789474
8         630    3.787879
9         106    3.754540
10        372    3.750000
11        195    3.729730
12        554    3.662500
13        311    3.622951

In [34]:
#Realizando a consulta
query ='''
SELECT
    a.author_id,
    a.author,
    z.avg_rating
FROM
    authors a
INNER JOIN(
SELECT
    b.author_id,
    AVG(t.avg_rating) AS avg_rating
FROM
    books b
INNER JOIN(
SELECT
    r.book_id,
    COUNT(r.book_id) AS "qtde_rating",
    AVG(r.rating) AS "avg_rating"
FROM
    ratings r
GROUP BY
    book_id
HAVING
    COUNT(r.book_id) > 50) t ON t.book_id = b.book_id
GROUP BY
    author_id
ORDER BY
    AVG(t.avg_rating) DESC) z ON z.author_id = a.author_id
ORDER BY
    z.avg_rating DESC
LIMIT 1;
'''

books_v = pd.io.sql.read_sql(query,con= engine)
books_v

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.283844

Conform visto acima, após aplicar o filtro de livros com mais de 50 classificações, e a partir disso calcular a média, nota-se que os 3 primeiros da listas possuem média de classificação com valores bem próximos. 

### Número médio de avaliações entre usuários que classificaram mais do que 50 livros.

In [35]:
ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [36]:
#Conjunto de dados rating (classificações)
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [37]:
#Conjunto de dados reviews (avaliações)
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [38]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


No enunciado é solicitado o cálculo do número médio de avaliações (review) dos usuários que realizaram mais de 50 classificações (rating). Porém há dois problemas: no conjunto de dados reviews não há o campo rating (notas) dos usuários; esse campo de encontra no conjunto de dados ratings. Outro ponto é que, agrupando o cojunto de dados reviews por username e contando os valores, nota-se que o número máximo de avaliações é 29 (número bem menor do que o pedido no enunciado, que é acima de 50). 

In [39]:
#Realizando a consulta
query='''
SELECT
    ra.username,
    COUNT(ra.username) AS "qtde_username",
    AVG(ra.rating) AS "avg_rating"
FROM
    ratings ra
GROUP BY
    ra.username
HAVING
    COUNT(ra.username) > 50
ORDER BY
    COUNT(ra.username) DESC
'''

book_user = pd.io.sql.read_sql(query, con=engine)
book_user

username  qtde_username  avg_rating
0          paul88             56    3.910714
1      martinadam             56    3.785714
2     sfitzgerald             55    3.836364
3       richard89             55    3.654545
4  jennifermiller             53    3.830189
5          xdavis             51    3.901961

Conforme visto acima e explicado anteriormente, tem-se acima os usuários que fizeram mais de 50 avaliações e suas respectivas médias. Nota-se que há empate entre primeiro e segundo lugares e terceiro e quarto lugares em quantidade de livros classificados. Percebe-se também que há usuários com média de avaliações maiores do que o segundo lugar, porém com quantidades menores de livros avaliados.